neonUtilities/R/zipsByProduct.R

```diff
@@ -338,27 +196,49 @@ zipsByProduct <- function(dpID, site="all", startdate=NA, enddate=NA, package="b
   }
   dir.create(filepath)
 
-  writeLines(paste("Downloading ", nrow(zip.urls)-1, " files", sep=""))
+  writeLines(paste("Downloading ", nrow(zip.urls), " files", sep=""))
   pb <- utils::txtProgressBar(style=3)
   utils::setTxtProgressBar(pb, 1/(nrow(zip.urls)-1))
-  # copy zip files into folder
-  for(i in 2:nrow(zip.urls)) {
-    zip_out <- paste(filepath, zip.urls$name[i], sep="/")
-    if(!file.exists(substr(zip_out, 1, nchar(zip_out)-4)) || !file.exists(zip_out)) {
-      downloader::download(zip.urls$URL[i], zip_out, 
-                           mode="wb", quiet=T)
-      utils::setTxtProgressBar(pb, i/(nrow(zip.urls)-1))
+
+  j <- 1
+  counter<- 1
+
+  while(j <= nrow(zip.urls)) {
+    counter <- counter + 1
+
+    if (counter > 3) {
+      cat(paste0("\nURL query ", zip.urls$name[j],
+                  " failed. The API or data product requested may be unavailable at this time; check data portal (data.neonscience.org/news) for possible outage alert."))
+      j <- j + 1
+    } else {
+      zip_out <- paste(filepath, zip.urls$name[j], sep="/")
+      if(!file.exists(substr(zip_out, 1, nchar(zip_out)-4)) || !file.exists(zip_out)) {
+        t <- tryCatch(
+          {
+            downloader::download(zip.urls$URL[j], zip_out,
+                                 mode="wb", quiet=T)
+          }, error = function(e) { e } )
+
+        if(inherits(t, "error")) {
+          writeLines("File could not be downloaded. URLs may have expired. Trying new URLs.")
+          zip.urls <- getZipUrls(tmp.files, month.urls=month.urls, avg=avg, package=package,
+                                 max.pub=max.pub, max.pub.site=max.pub.site, dpID=dpID, messages=messages) %>%
+            tidyr::drop_na()
+        } else {
+
+          messages[j] <- paste(zip.urls$name[j], "downloaded to", zip_out, sep=" ")
+          j = j + 1
+          counter <- 1
+        }
+      }
+      utils::setTxtProgressBar(pb, j/(nrow(zip.urls)-1))
     }
   }
+
   utils::setTxtProgressBar(pb, 1)
   close(pb)
 
-  if(load==F) {
```

Comment:
Don't drop this conditional. When load=T, the files are downloaded to a temporary directory, which gets deleted after processing, so it's confusing and weird for the end user to get a message giving them the file path to the download.